File: rle7.py
#! /bin/python

import struct
import io
import sys


def rle7_compress(data_in):
    """
    compress data using zero rle7 methread
    """
    in_len = len(data_in)
    data_out = io.BytesIO()

    cur_in = 0

    repeat = 0
    count = 1

    last = data_in[0]

    for cur_in in range(1, in_len):
        if data_in[cur_in] == last:
            if count != 1:
                # copy no repeat data
                count -= 1
                data_out.write(struct.pack('B', count - 2))

                while count != 1:
                    data_out.write(struct.pack('B', data_in[cur_in - count]))
                    count -= 1

            repeat += 1
        else:
            if repeat != 0:
                # write repeat data
                data_out.write(struct.pack('B', repeat | (1 << 7)))
                data_out.write(struct.pack('B', last))

                repeat = 0

            count += 1

            last = data_in[cur_in]

    if count != 1:
        # copy no repeat data
        count -= 1
        data_out.write(struct.pack('B', count - 1))

        while count != 0:
            data_out.write(struct.pack('B', data_in[cur_in - count]))
            count -= 1

    if repeat != 0:
        # write repeat data
        data_out.write(struct.pack('B', repeat | (1 << 7)))
        data_out.write(struct.pack('B', last))

    return data_out.getvalue()


def rle7_decompress(data_in, out_len):
    """
    decompress data using zero rle7 methread
    """
    data_out = io.BytesIO()

    cur_out = 0
    cur_in = 0
    cc = 0
    count = 0

    while cur_out < out_len:
        cc = data_in[cur_in]
        count = (cc & (~(1 << 7))) + 1

        if cc & (1 << 7):
            while count != 0:
                data_out.write(struct.pack('B', data_in[cur_in + 1]))
                cur_out += 1
                count -= 1

            cur_in += 2
        else:
            while count != 0:
                cur_in += 1
                data_out.write(struct.pack('B', data_in[cur_in]))
                cur_out += 1
                count -= 1

            cur_in += 1

    if cur_out != out_len:
        raise RuntimeError("decompress len mismatch")

    return data_out.getvalue()


def zero_rle_test():
    data = struct.pack('BBBBBBBBBBBBBBBBBBBBBBB', 0, 0, 0, 0,
                       0, 0, 1, 0, 0, 0, 5, 0, 6, 0, 0, 0, 1, 2, 3, 4, 0, 0, 0)
    print(data)
    print(rle7_compress(data))
    print(rle7_decompress(rle7_compress(data), len(data)))


if __name__ == "__main__":
    zero_rle_test()

File: rle7.c
#include <stdint.h>

#define SET_REPEAT(x) (x | (1 << 7))

uint32_t rle7_compress(uint8_t *out, const uint8_t *in, uint32_t in_len)
{
    uint32_t cur_out = 0;
    uint32_t cur_in  = 0;

    uint8_t  repeat  = 0;
    uint8_t  count   = 1;

    uint8_t  last    = *in;

    for (cur_in = 1; cur_in < in_len; cur_in++)
    {
        if (*(in + cur_in) == last)
        {
            if (1 != count)
            {
                // copy no repeat data
                *(out + cur_out++) = --count - 2;

                while (1 != count)
                {
                    *(out + cur_out++) = *(in + cur_in - count--);
                }
            }

            repeat++;
        }
        else
        {
            if (0 != repeat)
            {
                // write repeat data
                *(out + cur_out++) = SET_REPEAT(repeat);
                *(out + cur_out++) = last;

                repeat = 0;
            }

            count++;

            last = *(in + cur_in);
        }
    }

    if (1 != count)
    {
        // copy no repeat data
        *(out + cur_out++) = --count - 1;
        while (0 != count)
        {
            *(out + cur_out++) = *(in + cur_in - count--);
        }
    }

    if (0 != repeat)
    {
        // write repeat data
        *(out + cur_out++) = SET_REPEAT(repeat);
        *(out + cur_out++) = last;
    }

    return cur_out;
}

#define C_COUNT(x) (x & ~(1 << 7))
#define IS_REPEAT(x)  (x & (1 << 7))

uint32_t rle7_decompress(uint8_t *out, const uint8_t *in, uint32_t out_len)
{
    uint32_t cur_out = 0;
    uint32_t cur_in  = 0;
    uint8_t  cc      = 0;
    int8_t   count   = 0;

    while (cur_out < out_len)
    {
        cc    = *(in + cur_in);
        count = C_COUNT(cc) + 1;

        if (IS_REPEAT(cc))
        {
            while (count--)
            {
                *(out + cur_out++) = *(in + cur_in + 1);
            }
            cur_in += 2;
        }
        else
        {
            while (count--)
            {
                *(out + cur_out++) = *(in + ++cur_in);
            }
            cur_in++;
        }
    }

    // decompress len mismatch
    while (cur_out != out_len) {};

    return cur_in;
}
